lib/spack/spack/solver/asp.py

```diff
@@ -3367,15 +3369,19 @@ def virtual_on_edge(self, parent_node, provider_node, virtual):
         dependencies[0].update_virtuals((virtual,))
 
     def reorder_flags(self):
-        """Order compiler flags on specs in predefined order.
-
-        We order flags so that any node's flags will take priority over
-        those of its dependents.  That is, the deepest node in the DAG's
-        flags will appear last on the compile line, in the order they
-        were specified.
+        """For each spec, determine the order of compiler flags applied to it.
 
         The solver determines which flags are on nodes; this routine
-        imposes order afterwards.
+        imposes order afterwards. The order is:
+
+        1. Compiler flags should always come first
+        2. Command-line flags should always come last
+        3. For each source (requirements, compiler, command line, and dependents)
+           flags from that source should retain their order and grouping, e.g.
+           for `y cflags="-z -a"` "-z" and "-a" should never have any intervening
+           flags inserted, and should always appear in that order.
+        4. Flags from requirements, and dependents should be grouped and ordered
+           alphabetically (on a per-source basis, accounting for the prior point).
```

Comment:
@scheibelp why does this order dependent flags alphabetically when the prior logic used DAG order?  DAG order ensures that you get a deterministic ordering, but it also ensures that flags on a node override those inherited from its dependents. If you sort alphabetically you're not ensuring that.... it can change which node overrides which depending on the flags provided.  e.g.:

```
     A
    / \
   B   C
    \ /
     D
```

Any flags on D inherited from B and C should come after those inherited by A -- the order shouldn't depend on the name of the particular flag.  e.g., you don't want `-O3` to always take precedence over `-O2` regardless of whether A, B, or C specified it.

Can't we both preserve DAG order *and* preserve source grouping? I think you can add the `flag_group` to `node_flag_source` and preserve the `.index()`-based sorting I had below, but also preserve flag groups.

Reply:
I point out in the PR description that while we thought we did this, we don't. You can check out [19f5c6f](https://github.com/spack/spack/pull/41049/commits/19f5c6fe76ac04553d46f2658e19575c400c5a65) to test that yourself (new DAG ordering test with old logic).

Reply:
are you referring to `test_flag_order_and_grouping`?  It doesn't look like it's trying to enforce DAG order... which test do you want me to run on that?

Reply:
`test_two_dependents_flag_mixing`: this test only checks flag ordering for flags introduced by `depends_on` vs. any other form of mixing. If you run it 5-10 times it will fail some and succeed some.

Reply:
I removed `node_flag_source` and introduced `source` as a property of the `node_flag` attribute to properly associate introduced flags with their "origin" (i.e. whether they originated from a `depends_on`, or from a `require:`.

Right now, `depends_on` is the only origin that can be applied to one package from multiple sources; `require:` might do that in the future, so I added a `ConstraintOrigin` class. That could be removed while supporting everything we do now. Either way, the notion of origin needs to be plumbed through relevant methods in `asp.py` (e.g. for `package_dependencies_rules() -> condition() -> _get_condition_id() -> spec_clauses()`, we need to pass the origin info from the first to the last element, through all intermediates).